information_schema.views fixes for materialized views

findepi opened the pull request.

findepi commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorSmokeTest.java:
```diff
@@ -556,6 +556,20 @@ public void testMaterializedView()
                         "FROM\n" +
                         "  nation");
 
+        // information_schema.tables (no filtering on table_name so that ConnectorMetadata.listViews is exercised)
+        assertThat(query("SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = '" + schemaName + "'"))
+                .containsAll("VALUES (VARCHAR '" + viewName + "', VARCHAR 'BASE TABLE')");
+
+        // information_schema.views
+        assertThat(computeActual("SELECT table_name FROM information_schema.views WHERE table_schema = '" + schemaName + "'").getOnlyColumnAsSet())
+                .doesNotContain(viewName);
+        assertThat(query("SELECT table_name FROM information_schema.views WHERE table_schema = '" + schemaName + "' AND table_name = '" + viewName + "'"))
+                .returnsEmptyResult();
+
+        // materialized view-specific listings
+        assertThat(query("SELECT name FROM system.metadata.materialized_views WHERE catalog_name = '" + catalogName + "' AND schema_name = '" + schemaName + "'"))
+                .containsAll("VALUES VARCHAR '" + viewName + "'");
+
```
The new assertions being added here may be expensive.

let's see how much time the CI  spends on this -- this will be run for Iceberg only, since that's the only support materialized views

findepi replied:
TestIcebergConnectorSmokeTest.testMaterializedView; (took: 0.6 seconds)

losipiuk commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:
```diff
@@ -1104,7 +1104,10 @@ public void testMaterializedView()
 
         // view-specific listings
-        checkInformationSchemaViewsForMaterializedView(view.getSchemaName(), view.getObjectName());
+        assertThat(computeActual("SELECT table_name FROM information_schema.views WHERE table_schema = '" + view.getSchemaName() + "'").getOnlyColumnAsSet())
```
`Fix test expected behavior for informatio_schama.views for MV` typo in commit messsage

losipiuk replied:
2 actually

findepi replied:
thanks, fixed

findepi commented:
(just rebased)

kokosing commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -337,14 +337,6 @@ public void testShowCreateTable()
                         ")\\E");
```
> Despite a materialized view having `VIEW`
`table_type` in `information_schema.tables`, they are not supposed to
show up in `views`.

So where they should appear? Is this part of SQL standard?

findepi replied:
The commit message is wrong, will fix it.

we have system.metadata.materialized_views for them

findepi replied:
> Is this part of SQL standard?

afaik MVs are extension

findepi commented:
rebased to be on top of https://github.com/trinodb/trino/pull/17127 and fixed first commit message per https://github.com/trinodb/trino/pull/17444#discussion_r1191995040

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -2691,18 +2691,23 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
         for (String schemaName : listSchemas(session, optionalSchemaName)) {
             for (String tableName : metastore.getAllViews(schemaName)) {
-                tableNames.add(new SchemaTableName(schemaName, tableName));
+                SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+                if (!materializedViews.contains(schemaTableName)) {
+                    tableNames.add(schemaTableName);
+                }
```
new code cc @kokosing @huberty89 @raunaqmorarka

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -2691,18 +2691,23 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
             if (allViews.isPresent()) {
                 return allViews.get().stream()
                         .filter(view -> !isHiveSystemSchema(view.getSchemaName()))
+                        .filter(view -> !materializedViews.contains(view))
```
new code cc @kokosing @huberty89 @raunaqmorarka

huberty89 reviewed (approved):
LGTM

jklamer commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -2691,18 +2691,23 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
     @Override
     public List<SchemaTableName> listViews(ConnectorSession session, Optional<String> optionalSchemaName)
     {
+        Set<SchemaTableName> materializedViews = ImmutableSet.copyOf(listMaterializedViews(session, optionalSchemaName));
```
good news, I think we can remove similar logic from objectstore connector

findepi commented:
CI  https://github.com/trinodb/trino/issues/16277

kokosing commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorSmokeTest.java:
```diff
@@ -556,6 +556,20 @@ public void testMaterializedView()
 
+        // information_schema.tables (no filtering on table_name so that ConnectorMetadata.listViews is exercised)
+        assertThat(query("SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = '" + schemaName + "'"))
+                .containsAll("VALUES (VARCHAR '" + viewName + "', VARCHAR 'BASE TABLE')");
```
why view is marked as base table?

findepi replied:
Dictator's decision.

kokosing commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorSmokeTest.java:
```diff
@@ -556,6 +556,20 @@ public void testMaterializedView()
+                .returnsEmptyResult();
+
+        // materialized view-specific listings
+        assertThat(query("SELECT name FROM system.metadata.materialized_views WHERE catalog_name = '" + catalogName + "' AND schema_name = '" + schemaName + "'"))
```
are there any other columns that we should test?

findepi replied:
i chose `name` only because i focus on the listing ("is or isn't" there). we can revisit as needed

kokosing commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorSmokeTest.java:
```diff
@@ -556,6 +556,20 @@ public void testMaterializedView()
+                .containsAll("VALUES (VARCHAR '" + viewName + "', VARCHAR 'BASE TABLE')");
+
+        // information_schema.views
+        assertThat(computeActual("SELECT table_name FROM information_schema.views WHERE table_schema = '" + schemaName + "'").getOnlyColumnAsSet())
```
are there any other columns that we should test?

findepi replied:
as above

kokosing reviewed (approved):
comments

colebow commented:
Does this need release notes? @findepi

findepi commented:
i don't think so, Cole
